File: src/domain/amountEdit.ts
import { addMoney, type Halalas } from './money'
import { formatMoney } from './formatMoney'
import type { PersonAllocation, Settlement, Transaction } from './entities/types'

/**
 * تعديل مبلغ العملية — OVERRIDES §32 (قرار المالك 2026-09-15): «لو في عملية ثمنها 200 ودخل عدّلها
 * خلاها 180 مش مشكلة، والرقم يتحدث والمصروف الشامل يتحدث». يعلو على «المبلغ ما بيتعدلش» (spec/04).
 *
 * - المبلغ الجديد (`amountMinor`) هو اللي بيتحسب في **كل** حاجة.
 * - المبلغ الأصلي من المصدر بيتحفظ **مرة واحدة** في `originalAmountMinor` («أحتفظ بيه مستخبي»)،
 *   ومطابقة التكرار عند إعادة الاستيراد والدمج بتستعمله — فنفس سطر الكشف ما يتضافش تاني.
 * - المبلغ ما ينزلش تحت اللي متوزع على أشخاص أو متسوّى بيه دين (spec/06: لا رصيد سالب صامت).
 */
export class AmountEditError extends Error {}

/** المبلغ اللي مطابقة التكرار بتستعمله: الأصلي من المصدر لو اتعدل. */
export function sourceAmountMinor(txn: Pick<Transaction, 'amountMinor' | 'originalAmountMinor'>): Halalas {
  return txn.originalAmountMinor ?? txn.amountMinor
}

export function planAmountEdit(
  txn: Pick<Transaction, 'amountMinor' | 'originalAmountMinor'>,
  newAmountMinor: Halalas,
  allocations: readonly Pick<PersonAllocation, 'amountMinor'>[],
  settlements: readonly Pick<Settlement, 'amountMinor'>[],
): Pick<Transaction, 'amountMinor' | 'originalAmountMinor'> {
  if (!Number.isSafeInteger(newAmountMinor) || newAmountMinor <= 0) {
    throw new AmountEditError('اكتب مبلغ صحيح أكبر من صفر، زي 180 أو 180.50')
  }
  const allocated = allocations.reduce<Halalas>((sum, row) => addMoney(sum, row.amountMinor), 0)
  if (newAmountMinor < allocated) {
    throw new AmountEditError(`المبلغ أقل من اللي متوزع على أشخاص من العملية دي (${formatMoney(allocated)}). عدّل التوزيع الأول.`)
  }
  const settled = settlements.reduce<Halalas>((sum, row) => addMoney(sum, row.amountMinor), 0)
  if (newAmountMinor < settled) {
    throw new AmountEditError(`المبلغ أقل من اللي اتسدد بيه دين من العملية دي (${formatMoney(settled)}). عدّل التسوية الأول.`)
  }
  return { amountMinor: newAmountMinor, originalAmountMinor: sourceAmountMinor(txn) }
}

File: native-app/core/src/commonMain/kotlin/app/masroufy/core/AmountEdit.kt
package app.masroufy.core

/**
 * تعديل مبلغ العملية — نقل `src/domain/amountEdit.ts` (OVERRIDES §32).
 * الجديد بيتحسب في كل حاجة، والأصلي من المصدر بيتحفظ **مرة واحدة** لمطابقة التكرار،
 * والمبلغ ما ينزلش تحت اللي متوزع على أشخاص أو متسوّى بيه دين.
 */
class AmountEditError(message: String) : IllegalArgumentException(message)

data class AmountEdit(val amountMinor: Halalas, val originalAmountMinor: Halalas)

/** المبلغ اللي مطابقة التكرار بتستعمله: الأصلي من المصدر لو اتعدل. */
fun sourceAmountMinor(amountMinor: Halalas, originalAmountMinor: Halalas?): Halalas = originalAmountMinor ?: amountMinor

fun planAmountEdit(
    amountMinor: Halalas,
    originalAmountMinor: Halalas?,
    newAmountMinor: Halalas,
    allocated: List<Halalas>,
    settled: List<Halalas>,
): AmountEdit {
    if (newAmountMinor > MAX_SAFE_HALALAS || newAmountMinor <= 0) {
        throw AmountEditError(uiText(TextKey.AMOUNT_EDIT_INVALID))
    }
    val allocatedTotal = allocated.fold(0L) { sum, a -> addMoney(sum, a) }
    if (newAmountMinor < allocatedTotal) {
        throw AmountEditError(uiText(TextKey.AMOUNT_EDIT_BELOW_ALLOCATIONS, formatMoney(allocatedTotal)))
    }
    val settledTotal = settled.fold(0L) { sum, s -> addMoney(sum, s) }
    if (newAmountMinor < settledTotal) {
        throw AmountEditError(uiText(TextKey.AMOUNT_EDIT_BELOW_SETTLEMENTS, formatMoney(settledTotal)))
    }
    return AmountEdit(newAmountMinor, sourceAmountMinor(amountMinor, originalAmountMinor))
}
